src/services/codefixes/importFixes.ts

```diff
@@ -0,0 +1,474 @@
+                    // insert after any existing imports
+                    let lastModuleSpecifierEnd = -1;
+                    for (const moduleSpecifier of sourceFile.imports) {
+                        const end = moduleSpecifier.getEnd();
+                        if (!lastModuleSpecifierEnd || end > lastModuleSpecifierEnd) {
+                            lastModuleSpecifierEnd = end;
+                        }
+                    }
+                    const insertPos = lastModuleSpecifierEnd > 0 ? sourceFile.getLineEndOfPosition(lastModuleSpecifierEnd) : sourceFile.getStart();
+
+                    const getCanonicalFileName = createGetCanonicalFileName(useCaseSensitiveFileNames);
+                    const moduleSpecifierWithoutQuotes = stripQuotes(moduleSpecifier || getModuleSpecifierForNewImport());
+                    const importStatementText = isDefault
+                        ? `import ${name} from "${moduleSpecifierWithoutQuotes}"`
+                        : `import { ${name} } from "${moduleSpecifierWithoutQuotes}"`;
+
+                    // if this file doesn't have any import statements, insert an import statement and then insert a new line
+                    // between the only import statement and user code. Otherwise just insert the statement because chances
+                    // are there are already a new line seperating code and import statements.
+                    const newText = insertPos === sourceFile.getStart()
+                        ? `${importStatementText};${context.newLineCharacter}${context.newLineCharacter}`
+                        : `${context.newLineCharacter}${importStatementText};`;
+
+                    return createCodeAction(
+                        Diagnostics.Import_0_from_1,
+                        [name, `"${moduleSpecifierWithoutQuotes}"`],
+                        newText,
+                        { start: insertPos, length: 0 },
+                        sourceFile.fileName,
+                        "NewImport",
+                        moduleSpecifierWithoutQuotes
+                    );
+
+                    function getModuleSpecifierForNewImport() {
+                        const fileName = sourceFile.path;
+                        const moduleFileName = moduleSymbol.valueDeclaration.getSourceFile().path;
+                        const sourceDirectory = getDirectoryPath(fileName);
+                        const options = context.program.getCompilerOptions();
+
+                        return tryGetModuleNameFromAmbientModule() ||
+                            tryGetModuleNameFromBaseUrl() ||
+                            tryGetModuleNameFromRootDirs() ||
+                            tryGetModuleNameFromTypeRoots() ||
+                            tryGetModuleNameAsNodeModule() ||
+                            removeFileExtension(getRelativePath(moduleFileName, sourceDirectory));
+
+                        function tryGetModuleNameFromAmbientModule(): string {

```

Comment:
The module specifier is passed only when there is already an existing import declaration in this file, otherwise this is still needed.